### PR TITLE
Use getters and setters in field configs

### PR DESCRIPTION
**Pull request by ThomasBleijendaal**
In Entity Framwork Core, it is possible to create an owned object within an entity, which currently leads to some strange behaviour in Fluidity when using a custom repository. Also, when using properties from a sub-object in the list view, incorrect data is display. This PR fixes that by determining the getter and setter for each configured field.

To clarify:

Let's assume the following entity is defined:

```
class Person {
    string Name {get;set;}
    Company Company {get;set;}
}
[Owned]
class Company {
    string Name {get;set;}
}
```

When configuring the bootstrap for this Entity, one would use something like this:

```
treeConfig.AddCollection<Person>(p => p.Id, "Person", "People", "A collection of people", "icon-umb-users", "icon-umb-users", collectionConfig => {
    collectionConfig.SetNameProperty(p => p.Name);
    collectionConfig.SetViewMode(FluidityViewMode.List);

    collectionConfig.ListView(listViewConfig => {
        listViewConfig.AddField(p => p.Company.Name);
    });

    collectionConfig.Editor(editorConfig => {

        editorConfig.AddTab("General", tabConfig => {
            tabConfig.AddField(p => p.Company.Name);
        });
    });
});
```

When viewing these entities in the list view, both the Name and the Company columns will display the same names. When editing this entity, both the Name textbox and the Company Name textbox will display the same data. When saving it, the Person.Name column will have some data while Person.Company_Name will remain NULL in the database.

The cause of this issue comes from Fluidity using only the last element in the Expression used in the `AddField` methods. So for both properties, `Name` will be used, and will be read and written using `GetPropertyValue` and `SetPropertyValue`.

I have updated `FluidityPropertyConfig` the accept also a getter and a setter for its configured `propertyExp`. This getter and setter is created by the `GetterAndSetterHelper` which generates them as follows:

If the field `p => p.Company.Name` is configured, the getter `x => x.get_Company().get_Name()` and the setter `(x,y) => x.get_Company().set_Name(y)` is generated. The helper walks over the given expression until it reaches the root, so even more nested configuration are supported. This is especially useful in the ListView, when, for example, fields from related entities can be displayed. For example, when displaying a overview of tournaments, the name of the tournament's game is nice to have displayed.

`GetterAndSetterHelper` also creates a name for the field, so for the expression `x => x.Company.Owner.Name`, the name `CompanyOwnerName` is created. This name is used in the list view and in the editor, so the field will have a useful default heading and/or label. This configuration can be overridden just like before. 

When the `GetterAndSetterHelper` cannot determine a valid getter and setter, the current behaviour is used as fallback.

Let me know what you think and if I need to change anything to the code.

### CODE REVIEW

**Comment by mattbrailsford** [comment]
Hi @ThomasBleijendaal, this is actually pretty cool so thanks for spending the time on working this out. I think the approach is all good, but I think I would like it if the GetterAndSetterHelper could maybe work in a more generic way accepting a base LambdaExpression rather than the typed varient.

I get why you've done it this way as, if I'm reading correctly, you are using the types to know if you have reached the ends however I think this might be flawed, as what would happen if you have a hierarchical structure where you have the same entity type nested? 

I'm wondering if there is just more of a generic way to take a LambdaExperssion and break apart all the object getters regardless of the types?

**Comment by ThomasBleijendaal** [comment]
Hi Matt, I think converting the GetterAndSetterHelper is possible. I'm quite used to work with generics, so that's always my first option I try. But in this case, making it more generic by accepting a LambdaExpression is better. The helper should just work its way up the expression until it reaches the root. That also avoids its flaw in which it stops too early if you have the same entity type nested. 

It also solves that rather ugly conversion from a generic getter and setter to an object getter and setter (`(obj) => Getter.Invoke((TEntityType)obj) as object;`) since there will be only objects and no generic types.

File: src/Fluidity/Configuration/FluidityEditorFieldConfig`T.cs [hunk 1]
```diff
@@ -24,11 +25,25 @@ public class FluidityEditorFieldConfig<TEntityType, TValueType> : FluidityEditor
         /// <param name="propertyExpression">The property expression.</param>
         /// <param name="config">A configuration delegate.</param>
         public FluidityEditorFieldConfig(Expression<Func<TEntityType, TValueType>> propertyExpression, Action<FluidityEditorFieldConfig<TEntityType, TValueType>> config = null)
-            : base(propertyExpression)
         {
+            var getterAndSetter = GetterAndSetterHelper.Create(propertyExpression);
```
Comment (mattbrailsford) [inline src/Fluidity/Configuration/FluidityEditorFieldConfig`T.cs]: Now that GetterAndSetterHelper has been made more generic. Can this move back down to the base class?

File: src/Fluidity/Configuration/FluidityListViewFieldConfig`T.cs [hunk 2]
```diff
@@ -22,8 +23,22 @@ public class FluidityListViewFieldConfig<TEntityType, TValueType> : FluidityList
         /// <param name="propertyExpression">The property expression.</param>
         /// <param name="config">The configuration.</param>
         public FluidityListViewFieldConfig(Expression<Func<TEntityType, TValueType>> propertyExpression, Action<FluidityListViewFieldConfig<TEntityType, TValueType>> config = null)
-            : base (propertyExpression)
         {
+            var getterAndSetter = GetterAndSetterHelper.Create(propertyExpression);
```
Comment (mattbrailsford) [inline src/Fluidity/Configuration/FluidityListViewFieldConfig`T.cs]: Same as above. To move back down to the base class.

**Review by mattbrailsford — CHANGES_REQUESTED** [review]
Just a couple of tweaks now that the GetterAndSetterHelper has been made more generic. 

Fantastic work 👍

**Comment by ThomasBleijendaal** [comment]
Yes, it's better to have the constructor of `FluidityPropertyConfig` determine whether it can use the getter and setter.

**Comment by ThomasBleijendaal** [comment]
~I think I have to make one more change to this PR before it's completely ready: It's not necessary to completely determine a getter and a setter for the `FluidityListViewFieldConfig`. If I only determine a valid getter for the list view config, expressions containing some logic can be used in the list view (like counting the number of sub-entities, combining multiple sub-entities using `string.Join()`, etc).~

I see that this has been disallowed by `LambdaExpressionExtensions`, and modifying this will impact a lot more functionality. It's probably best to skip this or move this to another PR. But, it should be possible to create getters from any expression now, and determine it's name using the new helper. Those two things is probably enough for `FluidityListViewFieldConfig` to fully work, and skip getting the `PropertyInfo` from the lambda expression.

**Comment by ThomasBleijendaal** [comment]
Hi @mattbrailsford, I have made the changes you requested in your review, and made a small change which will make invoking the getters and setters faster by switching to `Invoke()` since it's not required to use `DynamicInvoke()`. Is it possible to create a new nuget release after this PR has been merged? I currently have various projects based upon Fluidity, and this change would remove a lot of workarounds.

**Comment by mattbrailsford** [comment]
Hi @ThomasBleijendaal, great work.

I'll try and take a look tomorrow and see if I can't get it merged in. I need to review another PR as well so hopefully if I can get those done I can get them merged and a new release out the door.

Thanks for the hard work on this one.

**Comment by mattbrailsford** [comment]
@ThomasBleijendaal  just wanted to say, I've not forgotten about this PR, just things are mega busy atm. Hoping to get some time to test next week. I'll keep you posted and thanks again for the work on this. I will get it merged in ASAP.

**Review by mattbrailsford — APPROVED** [review]

**Pull request merged**